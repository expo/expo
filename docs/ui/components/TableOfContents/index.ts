export * from './TableOfContents';
export * from './TableOfContentsLink';
