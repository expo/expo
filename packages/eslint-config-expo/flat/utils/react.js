const pluginReact = require('eslint-plugin-react');
const pluginReactHooks = require('eslint-plugin-react-hooks');

module.exports = [
  {
    plugins: {
      react: pluginReact,
      'react-hooks': pluginReactHooks,
    },

    languageOptions: {
      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...pluginReactHooks.configs.recommended.rules,
      'react/display-name': 'warn',
      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',
      'react/no-danger-with-children': 'warn',
      'react/no-deprecated': 'warn',
      'react/no-direct-mutation-state': 'warn',

      'react/no-string-refs': [
        'warn',
        {
          noTemplateLiterals: true,
        },
      ],

      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/require-render-return': 'warn',
    },
  },
];
