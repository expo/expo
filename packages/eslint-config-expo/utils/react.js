module.exports = {
  parserOptions: { ecmaFeatures: { jsx: true } },
  plugins: ['react'],
  extends: ['plugin:react-hooks/recommended'],
  rules: {
    'react/display-name': 'warn',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-uses-react': 'warn',
    'react/jsx-uses-vars': 'warn',
    'react/no-danger-with-children': 'warn',
    'react/no-deprecated': 'warn',
    'react/no-direct-mutation-state': 'warn',
    'react/no-string-refs': ['warn', { noTemplateLiterals: true }],
    'react/no-this-in-sfc': 'warn',
    'react/no-unknown-property': 'warn',
    'react/require-render-return': 'warn',
  },
  settings: {
    react: { version: 'detect' },
  },
};
