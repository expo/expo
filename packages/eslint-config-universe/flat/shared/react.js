const { defineConfig } = require('eslint/config');
const react = require('eslint-plugin-react');
const hooksPlugin = require('eslint-plugin-react-hooks');

module.exports = defineConfig([
  {
    plugins: {
      react,
      'react-hooks': hooksPlugin,
    },

    languageOptions: {
      // TODO: Const is a reserved word
      // ecmaVersion: 5,
      // sourceType: 'script',

      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...hooksPlugin.configs.recommended.rules,
      'react/jsx-boolean-value': ['warn', 'never'],

      'react/jsx-closing-bracket-location': [
        'warn',
        {
          nonEmpty: 'after-props',
          selfClosing: 'tag-aligned',
        },
      ],

      'react/jsx-curly-brace-presence': ['warn', 'never'],

      'react/jsx-curly-spacing': [
        'warn',
        {
          when: 'never',
        },
      ],

      'react/jsx-equals-spacing': ['warn', 'never'],
      'react/jsx-first-prop-new-line': ['warn', 'multiline'],
      'react/jsx-fragments': ['warn', 'syntax'],
      'react/jsx-indent': ['warn', 2],
      'react/jsx-indent-props': ['warn', 2],

      'react/jsx-no-bind': [
        'warn',
        {
          allowArrowFunctions: true,
          allowFunctions: true,
        },
      ],

      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-props-no-multi-spaces': 'warn',
      'react/jsx-tag-spacing': 'warn',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',

      'react/jsx-wrap-multilines': [
        'warn',
        {
          declaration: false,
          assignment: false,
          return: true,
          arrow: true,
        },
      ],

      'react/no-access-state-in-setstate': 'warn',
      'react/no-did-mount-set-state': 'warn',
      'react/no-did-update-set-state': 'warn',
      'react/no-direct-mutation-state': 'warn',
      'react/no-redundant-should-component-update': 'warn',
      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/no-will-update-set-state': 'warn',
      'react/require-render-return': 'warn',
      'react/self-closing-comp': 'warn',
      'react-hooks/rules-of-hooks': 'error',
      'react-hooks/exhaustive-deps': 'off',
    },
  },
]);
