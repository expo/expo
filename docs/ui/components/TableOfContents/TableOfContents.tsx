import { Button, mergeClasses } from '@expo/styleguide';
import { ArrowCircleUpIcon } from '@expo/styleguide-icons/outline/ArrowCircleUpIcon';
import { LayoutAlt03Icon } from '@expo/styleguide-icons/outline/LayoutAlt03Icon';
import {
  PropsWithChildren,
  RefObject,
  MouseEvent,
  forwardRef,
  useState,
  useRef,
  useImperativeHandle,
  useEffect,
} from 'react';

import { BASE_HEADING_LEVEL, Heading } from '~/common/headingManager';
import { prefersReducedMotion } from '~/common/window';
import { HeadingManagerProps, HeadingsContext } from '~/common/withHeadingManager';
import { ScrollContainer } from '~/components/ScrollContainer';
import { TableOfContentsLink } from '~/ui/components/TableOfContents';
import { CALLOUT } from '~/ui/components/Text';

const UPPER_SCROLL_LIMIT_FACTOR = 1 / 4;
const LOWER_SCROLL_LIMIT_FACTOR = 7 / 8;
const ACTIVE_ITEM_OFFSET_FACTOR = 1 / 20;

export type TableOfContentsProps = PropsWithChildren<{
  maxNestingDepth?: number;
  selfRef?: RefObject<ScrollContainer>;
  contentRef?: RefObject<ScrollContainer>;
}>;

export type TableOfContentsHandles = {
  handleContentScroll?: (contentScrollPosition: number) => void;
};

export const TableOfContents = forwardRef<
  TableOfContentsHandles,
  HeadingManagerProps & TableOfContentsProps
>(({ headingManager: { headings }, contentRef, selfRef, maxNestingDepth = 4 }, ref) => {
  const [activeSlug, setActiveSlug] = useState<string | null>(null);
  const [showScrollTop, setShowScrollTop] = useState(false);
  const [reducedMotion, setReducedMotion] = useState(false);

  const slugScrollingTo = useRef<string | null>(null);
  const activeItemRef = useRef<HTMLAnchorElement | null>(null);

  useEffect(function didMount() {
    if (contentRef?.current) {
      handleContentScroll(contentRef.current.getScrollTop());
    }
    setReducedMotion(prefersReducedMotion());
  }, []);

  useEffect(
    function didActiveSlugChanged() {
      updateSelfScroll();
    },
    [activeSlug]
  );

  useImperativeHandle(ref, () => ({ handleContentScroll }), []);

  function handleContentScroll(contentScrollPosition: number) {
    for (const { ref, slug } of headings) {
      if (!ref || !ref.current) continue;

      setShowScrollTop(contentScrollPosition > 120);

      if (
        ref.current.offsetTop >=
          contentScrollPosition + window.innerHeight * ACTIVE_ITEM_OFFSET_FACTOR &&
        ref.current.offsetTop <= contentScrollPosition + window.innerHeight / 2
      ) {
        if (slug !== activeSlug) {
          if (slug === slugScrollingTo.current) {
            slugScrollingTo.current = null;
          }
          setActiveSlug(slug);
          updateSelfScroll();
        }
        return;
      }
    }
  }

  function updateSelfScroll() {
    const selfScroll = selfRef?.current?.getScrollRef().current;
    const activeItemPos = activeItemRef.current?.offsetTop;

    if (!selfScroll || !activeItemPos || slugScrollingTo.current) {
      return;
    }

    const { scrollTop } = selfScroll;
    const upperThreshold = window.innerHeight * UPPER_SCROLL_LIMIT_FACTOR;
    const lowerThreshold = window.innerHeight * LOWER_SCROLL_LIMIT_FACTOR;

    if (activeItemPos < scrollTop + upperThreshold) {
      selfScroll.scrollTo({
        behavior: reducedMotion ? 'instant' : 'auto',
        top: Math.max(0, activeItemPos - upperThreshold),
      });
    } else if (activeItemPos > scrollTop + lowerThreshold) {
      selfScroll.scrollTo({
        behavior: reducedMotion ? 'instant' : 'auto',
        top: activeItemPos - lowerThreshold,
      });
    }
  }

  function handleLinkClick(event: MouseEvent, { slug, ref, type }: Heading) {
    event.preventDefault();

    slugScrollingTo.current = slug;

    const scrollOffset = type === 'inlineCode' ? 50 : 26;

    contentRef?.current?.getScrollRef().current?.scrollTo({
      behavior: reducedMotion ? 'instant' : 'smooth',
      top: ref.current?.offsetTop - window.innerHeight * ACTIVE_ITEM_OFFSET_FACTOR - scrollOffset,
    });

    if (history?.replaceState) {
      history.replaceState(history.state, '', '#' + slug);
    }
  }

  function handleTopClick(event: MouseEvent) {
    event.preventDefault();

    contentRef?.current?.getScrollRef().current?.scrollTo({
      behavior: reducedMotion ? 'instant' : 'smooth',
      top: 0,
    });

    if (history?.replaceState) {
      history.replaceState(history.state, '', ' ');
    }
  }

  const displayedHeadings = headings.filter(
    head =>
      head.level <= BASE_HEADING_LEVEL + maxNestingDepth && head.title.toLowerCase() !== 'see also'
  );

  return (
    <nav className="w-[280px] px-6 pb-10 pt-14" data-sidebar>
      <CALLOUT
        weight="medium"
        className="absolute z-10 -mt-14 mb-2 flex min-h-[32px] w-[248px] select-none items-center gap-2 bg-default pb-2 pt-4">
        <LayoutAlt03Icon className="icon-sm" /> On this page
        <Button
          theme="quaternary"
          size="xs"
          className={mergeClasses(
            'ml-auto mr-2 px-2 transition-opacity duration-300',
            !showScrollTop && 'pointer-events-none opacity-0'
          )}
          onClick={handleTopClick}>
          <ArrowCircleUpIcon className="icon-sm text-icon-secondary" aria-label="Scroll to top" />
        </Button>
      </CALLOUT>
      {displayedHeadings.map(heading => {
        const isActive = heading.slug === activeSlug;
        return (
          <TableOfContentsLink
            key={heading.slug}
            heading={heading}
            onClick={event => handleLinkClick(event, heading)}
            isActive={isActive}
            ref={isActive ? activeItemRef : undefined}
            shortenCode
          />
        );
      })}
    </nav>
  );
});

export const TableOfContentsWithManager = forwardRef<TableOfContentsHandles, TableOfContentsProps>(
  (props, ref) => {
    return (
      <HeadingsContext.Consumer>
        {headingManager => (
          <TableOfContents ref={ref} headingManager={headingManager!} {...props} />
        )}
      </HeadingsContext.Consumer>
    );
  }
);
