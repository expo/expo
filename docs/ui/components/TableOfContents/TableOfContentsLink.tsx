import { mergeClasses } from '@expo/styleguide';
import Link from 'next/link';
import { forwardRef, useState, type MouseEvent } from 'react';

import { BASE_HEADING_LEVEL, Heading, HeadingType } from '~/common/headingManager';
import { MONOSPACE, CALLOUT, FOOTNOTE } from '~/ui/components/Text';
import * as Tooltip from '~/ui/components/Tooltip';

type SidebarLinkProps = {
  heading: Heading;
  isActive: boolean;
  shortenCode: boolean;
  onClick: (event: MouseEvent<HTMLAnchorElement>) => void;
};

export const TableOfContentsLink = forwardRef<HTMLAnchorElement, SidebarLinkProps>(
  ({ heading: { slug, level, title, type, tags }, isActive, shortenCode, onClick }, ref) => {
    const [tooltipVisible, setTooltipVisible] = useState(false);

    // preset for monospace, tail ellipsis, and removing extra bits like details of function signatures
    const isCode = type === HeadingType.InlineCode;
    // preset for monospace, tail ellipsis, don't touch the title otherwise
    const isCodeOrFilePath = isCode || type === HeadingType.CodeFilePath;

    const TitleElement = isCodeOrFilePath ? MONOSPACE : CALLOUT;
    const displayTitle = shortenCode && isCode ? trimCodedTitle(title) : title;
    const isDeprecated = tags && tags.length > 0 ? tags.find(tag => tag === 'deprecated') : null;

    function onMouseOver(event: MouseEvent<HTMLAnchorElement>) {
      setTooltipVisible(isOverflowing(event.currentTarget));
    }

    function onMouseOut() {
      setTooltipVisible(false);
    }

    return (
      <Tooltip.Root open={tooltipVisible}>
        <Tooltip.Trigger asChild>
          <Link
            ref={ref}
            onMouseOver={isCode ? onMouseOver : undefined}
            onMouseOut={isCode ? onMouseOut : undefined}
            href={'#' + slug}
            onClick={onClick}
            className={mergeClasses(
              'mb-1.5 flex items-center justify-between truncate !text-pretty',
              convertToIndentClass(level - BASE_HEADING_LEVEL),
              'focus-visible:relative focus-visible:z-10'
            )}>
            <TitleElement
              className={mergeClasses(
                'w-full !text-secondary hocus:!text-link',
                isCodeOrFilePath && 'truncate !text-2xs',
                isActive && '!text-link',
                isDeprecated && 'line-through opacity-80'
              )}>
              {displayTitle}
            </TitleElement>
          </Link>
        </Tooltip.Trigger>
        <Tooltip.Content
          side="bottom"
          align="start"
          collisionPadding={{
            right: 22,
          }}>
          <FOOTNOTE tag={isCode ? 'code' : undefined}>{displayTitle}</FOOTNOTE>
        </Tooltip.Content>
      </Tooltip.Root>
    );
  }
);

/**
 * Replaces `Module.someFunction<T>(arguments: argType)` with `someFunction()`
 */
function trimCodedTitle(str: string) {
  if (!str.includes('...')) {
    const dotIdx = str.indexOf('.');
    if (dotIdx > 0) str = str.substring(dotIdx + 1);
  }

  str = str.replace(/<.+>/g, '');

  const parIdx = str.indexOf('(');
  if (parIdx > 0) str = str.substring(0, parIdx + 1) + ')';

  return str;
}

/**
 * Determines if element is overflowing (children width exceeds container width).
 * @param {HTMLElement} el HTML element to check
 */
function isOverflowing(el: HTMLElement) {
  if (!el || !el.children) {
    return false;
  }

  const childrenWidth = Array.from(el.children).reduce((sum, child) => sum + child.scrollWidth, 0);
  const indent = parseInt(window.getComputedStyle(el).paddingLeft, 10);
  return childrenWidth > 220 && childrenWidth >= el.scrollWidth - indent;
}

function convertToIndentClass(spacing: number) {
  switch (spacing) {
    case 1:
      return 'pl-3';
    case 2:
      return 'pl-6';
    case 3:
      return 'pl-9';
    case 4:
      return 'pl-12';
    case 5:
      return 'pl-15';
    default:
      return '';
  }
}
