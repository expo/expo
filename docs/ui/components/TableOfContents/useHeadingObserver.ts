import { useEffect, useRef, useState } from 'react';

export type HeadingEntry = {
  id: string;
  element: HTMLHeadingElement;
};

/**
 * Retrieve all headings matching the selector within the document.
 * This will search for all elements matching the heading tags, and generate a selector string.
 * The string can be used to properly intialize the intersection observer.
 * Currently, only headings with either an `[id="<id>"]` or `[data-id="<id>"]` are supported.
 */
export function useHeadingsObserver(tags = 'h2,h3') {
  const observerRef = useRef<IntersectionObserver>();
  const [headings, setHeadings] = useState<HeadingEntry[]>([]);
  const [activeId, setActiveId] = useState<string>();

  useEffect(
    function didMount() {
      const headings = Array.from(document.querySelectorAll<HTMLHeadingElement>(tags)).map(
        element => ({ element, id: getHeadingId(element) })
      );

      function onObserve(entries: IntersectionObserverEntry[]) {
        const entry = getActiveEntry(entries);
        if (entry) {
          setActiveId(getHeadingId(entry.target as HTMLHeadingElement));
        }
      }

      observerRef.current = new IntersectionObserver(onObserve, {
        // TODO(cedric): make sure these margins are tweaked properly with the new heading components
        rootMargin: '-25% 0px -50% 0px',
      });

      setHeadings(headings);
      headings.forEach(heading => observerRef.current?.observe(heading.element));

      return function didUnmount() {
        observerRef.current?.disconnect();
      };
    },
    [tags]
  );

  return { headings, activeId };
}

/**
 * Get the unique identifier of the heading element.
 * This could either be `[id="<id>"]` or `[data-id="<id>"]`.
 */
function getHeadingId(heading: HTMLHeadingElement): string {
  return heading.id || heading.dataset.id || '';
}

/**
 * Find the most probable observer entry that should be considered "active".
 * If there are more than one, the upper heading (by offsetTop) is returned.
 */
function getActiveEntry(entries: IntersectionObserverEntry[]): IntersectionObserverEntry | null {
  const visible = entries.filter(entry => entry.isIntersecting);

  if (visible.length <= 0) {
    return null;
  } else if (visible.length === 1) {
    return visible[0];
  }

  const sorted = visible.sort(
    (a, b) =>
      (a.target as HTMLHeadingElement).offsetTop - (b.target as HTMLHeadingElement).offsetTop
  );

  return sorted[0];
}
