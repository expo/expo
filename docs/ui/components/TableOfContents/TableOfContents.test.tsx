import { spacing } from '@expo/styleguide';
import { render, screen } from '@testing-library/react';
import React from 'react';

import { getHeadingIndent, getHeadingInfo } from './TableOfContents';

describe(getHeadingIndent, () => {
  const paddingFactor = spacing[2];

  // This shouldn't be included in the table of contents, but it's nice to test anyways
  it('returns no padding for h1 heading', () => {
    expect(getHeadingIndent(makeHeading('h1'))).toHaveProperty('paddingLeft', 0);
  });

  it('returns no padding for h2 heading', () => {
    expect(getHeadingIndent(makeHeading('h2'))).toHaveProperty('paddingLeft', 0);
  });

  it('returns 1x padding for h3 heading', () => {
    expect(getHeadingIndent(makeHeading('h3'))).toHaveProperty('paddingLeft', paddingFactor);
  });

  it('returns 2x padding for h4 heading', () => {
    expect(getHeadingIndent(makeHeading('h4'))).toHaveProperty('paddingLeft', paddingFactor * 2);
  });

  it('returns 3x padding for h5 heading', () => {
    expect(getHeadingIndent(makeHeading('h5'))).toHaveProperty('paddingLeft', paddingFactor * 3);
  });

  it('returns 4x padding for h6 heading', () => {
    expect(getHeadingIndent(makeHeading('h6'))).toHaveProperty('paddingLeft', paddingFactor * 4);
  });
});

describe(getHeadingInfo, () => {
  it('returns normal text from h1 heading', () => {
    expect(getHeadingInfo(makeHeading('h1', 'Hello'))).toMatchObject({
      type: 'text',
      text: 'Hello',
    });
  });

  it('returns normal text from h2 heading', () => {
    expect(getHeadingInfo(makeHeading('h2', 'Hello World'))).toMatchObject({
      type: 'text',
      text: 'Hello World',
    });
  });

  it('returns normal text from h3 heading with platform specification', () => {
    expect(getHeadingInfo(makeHeading('h3', 'Cool stuff (Android only)'))).toMatchObject({
      type: 'text',
      text: 'Cool stuff (Android only)',
    });
  });

  it('returns code text from h4 heading with function name', () => {
    expect(getHeadingInfo(makeHeading('h4', 'getCoolStuffAsync()'))).toMatchObject({
      type: 'code',
      text: 'getCoolStuffAsync',
    });
  });

  it('returns code text from h5 heading with function name and args', () => {
    expect(getHeadingInfo(makeHeading('h5', 'getTransformAsync(input: string)'))).toMatchObject({
      type: 'code',
      text: 'getTransformAsync',
    });
  });

  it('returns code text from h6 heading with function name, args, and return type', () => {
    expect(
      getHeadingInfo(makeHeading('h6', 'getTransformAsync(input: string): Promise<string>'))
    ).toMatchObject({
      type: 'code',
      text: 'getTransformAsync',
    });
  });
});

function makeHeading(Tag: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6', text = 'Hello World') {
  render(<Tag data-testid="heading">{text}</Tag>);
  return screen.getByTestId('heading') as HTMLHeadingElement;
}
