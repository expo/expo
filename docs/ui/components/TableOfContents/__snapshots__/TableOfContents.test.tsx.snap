// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableOfContents correctly matches snapshot 1`] = `
<div>
  <nav
    class="w-[280px] px-6 pb-10 pt-14"
    data-sidebar="true"
  >
    <p
      class="text-default text-[14px] leading-[1.5715] tracking-[-0.006rem] font-medium absolute z-10 -mt-14 mb-2 flex min-h-[32px] w-[248px] select-none items-center gap-2 bg-default pb-2 pt-4"
      data-text="true"
    >
      <svg
        class="text-icon-default translate-z icon-sm"
        fill="none"
        role="img"
        stroke="currentColor"
        viewBox="0 0 24 24"
      >
        <g
          id="layout-alt-03-outline-icon"
        >
          <path
            d="M17.5 17H6.5M17.5 13H6.5M3 9H21M7.8 3H16.2C17.8802 3 18.7202 3 19.362 3.32698C19.9265 3.6146 20.3854 4.07354 20.673 4.63803C21 5.27976 21 6.11984 21 7.8V16.2C21 17.8802 21 18.7202 20.673 19.362C20.3854 19.9265 19.9265 20.3854 19.362 20.673C18.7202 21 17.8802 21 16.2 21H7.8C6.11984 21 5.27976 21 4.63803 20.673C4.07354 20.3854 3.6146 19.9265 3.32698 19.362C3 18.7202 3 17.8802 3 16.2V7.8C3 6.11984 3 5.27976 3.32698 4.63803C3.6146 4.07354 4.07354 3.6146 4.63803 3.32698C5.27976 3 6.11984 3 7.8 3Z"
            id="Icon"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </g>
      </svg>
       On this page
      <button
        class="cursor-pointer inline-flex border border-solid rounded-md font-medium items-center whitespace-nowrap gap-1.5 h-8 text-3xs border-button-quaternary bg-button-quaternary text-button-quaternary shadow-none hocus:bg-button-quaternary-hover active:scale-98 ml-auto mr-2 px-2 transition-opacity duration-300 pointer-events-none opacity-0"
        type="button"
      >
        <span
          class="flex self-center text-inherit leading-none"
        >
          <svg
            aria-label="Scroll to top"
            class="translate-z icon-sm text-icon-secondary"
            fill="none"
            role="img"
            stroke="currentColor"
            viewBox="0 0 24 24"
          >
            <g
              id="arrow-circle-up-outline-icon"
            >
              <path
                d="M16 12L12 8M12 8L8 12M12 8V16M22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12Z"
                id="Icon"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </g>
          </svg>
        </span>
      </button>
    </p>
    <a
      class="mb-1.5 flex items-center justify-between truncate !text-pretty focus-visible:relative focus-visible:z-10"
      data-state="closed"
      href="#base-level-heading"
    >
      <p
        class="text-default font-normal text-[14px] leading-[1.5715] tracking-[-0.006rem] w-full !text-secondary hocus:!text-link"
        data-text="true"
      >
        Base level heading
      </p>
    </a>
    <a
      class="mb-1.5 flex items-center justify-between truncate !text-pretty pl-3 focus-visible:relative focus-visible:z-10"
      data-state="closed"
      href="#level-3-subheading"
    >
      <p
        class="text-default font-normal text-[14px] leading-[1.5715] tracking-[-0.006rem] w-full !text-secondary hocus:!text-link"
        data-text="true"
      >
        Level 3 subheading
      </p>
    </a>
    <a
      class="mb-1.5 flex items-center justify-between truncate !text-pretty pl-3 focus-visible:relative focus-visible:z-10"
      data-state="closed"
      href="#code-heading-depth-1"
    >
      <code
        class="leading-[1.6154] text-default w-full !text-secondary hocus:!text-link truncate !text-2xs"
        data-text="true"
      >
        Code heading depth 1
      </code>
    </a>
  </nav>
</div>
`;
